### PR TITLE
Compatibility with BBEdit

### PR DESCRIPTION
**Pull request by mixio**
Thank you for your good work on PLS.

I have been testing it with BBEdit which in its latest version supports language servers.

https://www.barebones.com/support/bbedit/lsp-notes.html

I came over 2 issues I had to get around and a feature request.

Issue 1. 
    
    PLS didn't support the 'shutdown' and 'exit' messages has required by the protocol specification.

    https://microsoft.github.io/language-server-protocol/specification#shutdown
    https://microsoft.github.io/language-server-protocol/specification#exit
    
    Has a consequence BBEdit was hanging when quitting or shutting down the server.
    
    Thanks to the clean architecture of PLS, it was easy to implement support for those message.
    
Issue 2. 
    
    I couldn't figure out how to make BBEdit correctly answer to the 'workspace/configuration' request.
    I temporarily short-circuited the response and hard-coded a default and reported the issue to the BBEdit support team.
    
Feature request
    
    To figure out what was going on in the conversation between BBEdit and PLS I implemented a primitive form of logging.
    It would be great if some more configurable form of logging was implemented in PLS.
    
With this provisional fixes, BBEdit now starts and stops the server correctly and appears to take advantage of the functionality provided by PLS.

### CODE REVIEW

**Comment by mixio** [comment]
The above Issue 2 could be solved with the help of BBEdit's developer.
BBEdit was expecting a 'scopeUri' parameter in its 'workspace/configuration' request even though it's optional per the specification.

https://microsoft.github.io/language-server-protocol/specification#workspace_configuration

Implementing support for the 'scopeUri' parameter in PLS might be a useful addition.

**Comment by FractalBoy** [comment]
I'll try to review this today or tomorrow.

File: server/lib/JJ.pm [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+Simple debugging helper that logs in BBEdits Logs directory.
+
+=cut
+
+sub jjlog {
+    my ( $sender, $object ) = @_;
+    my $log_file_path =
+        "~/Library/Containers/com.barebones.bbedit/Data/Library/Logs/BBEdit/perl-PLS.log";
+    open( FH, '>>', glob( $log_file_path ) ) or die $!;
```
Comment (FractalBoy) [inline server/lib/JJ.pm:24]: Bareword filehandles are not acceptable in this codebase. You must use a lexical filehandle.

Additionally, my style dictates that open should not have parentheses.

```suggestion
    open my $fh, '>>', glob($log_file_path) or die $!;
```

File: server/lib/JJ.pm [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+JJ
+
+=head1 DESCRIPTION
+
+Simple debugging helper that logs in BBEdits Logs directory.
+
+=cut
+
+sub jjlog {
```
Comment (FractalBoy) [inline server/lib/JJ.pm:20]: This file must be tidied using these perltidy args:

`-csc -gnu -l200`

File: server/lib/PLS/Server.pm [hunk 3]
```diff
@@ -133,6 +136,8 @@ sub run
                 my $json = substr $$buffref, 0, $size, '';
                 $size = 0;
 
+                JJ::jjlog('_C', $json);
```
Comment (FractalBoy) [inline server/lib/PLS/Server.pm:139]: What is _C?

File: server/lib/PLS/Server.pm [hunk 4]
```diff
@@ -209,6 +214,9 @@ sub send_message
 
     return if (not blessed($message) or not $message->isa('PLS::Server::Message'));
     my $json   = $message->serialize();
+
+    JJ::jjlog('S_', $json);
```
Comment (FractalBoy) [inline server/lib/PLS/Server.pm:218]: What is S_? Why is the underscore after the S here but before the C above?

File: server/lib/PLS/Server/Method/TextDocument.pm [hunk 5]
```diff
@@ -126,6 +127,10 @@ sub get_request
     {
         return PLS::Server::Request::TextDocument::Completion->new($request);
     }
+    if ($method eq 'publishDiagnostics')
```
Comment (FractalBoy) [inline server/lib/PLS/Server/Method/TextDocument.pm:130]: Why did you add this? publishDiagnostics is a request from the server to the client - the client will never send this request, so it's not needed here.

File: server/lib/PLS/Server/Request/Exit.pm [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+=head1 DESCRIPTION
+
+This is a notification message from the client to the server requesting
+that the server exits.
+
+=cut
+
+sub service
+{
+    exit 1;
```
Comment (FractalBoy) [inline server/lib/PLS/Server/Request/Exit.pm:23]: Why is this `exit 1`? The Exit request is a graceful exit (or should be) - this should be `exit 0`.

Comment (FractalBoy) [reply server/lib/PLS/Server/Request/Exit.pm:23]: Correction - the exit code should be 0 if Shutdown has been sent, 1 otherwise.

>A notification to ask the server to exit its process. The server should exit with success code 0 if the shutdown request has been received before; otherwise with error code 1.

File: server/lib/PLS/Server/Request/Shutdown.pm [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+This is a notification message from the client to the server requesting
+that the server shuts down.
+
+=cut
+
+sub service
+{
+    my ($self) = @_;
+
+    return PLS::Server::Response::Shutdown->new($self);
```
Comment (FractalBoy) [inline server/lib/PLS/Server/Request/Shutdown.pm:25]: This or the Response should probably set a variable in `PLS::Server::State` indicating that shutdown has been requested.

File: server/t/packets/shutdown.json [hunk 8]
```diff
@@ -0,0 +1,6 @@
+{
+    "jsonrpc": "2.0",
+    "id": null,
```
Comment (FractalBoy) [inline server/t/packets/shutdown.json:3]: This needs an id otherwise the response won't have an id.

**Review by FractalBoy — CHANGES_REQUESTED** [review]

File: server/lib/JJ.pm [hunk 9]
```diff
@@ -0,0 +1,31 @@
+package JJ;
```
[the remark this answers is not in the document]
Comment (mixio) [reply server/lib/JJ.pm:1]: Thank you for reviewing this PR.

Obviously the logging code was not meant to be pulled "as is". So please ignore it.
I nonetheless think a logging facility is of great help when debugging an lsp server and will leave it to you to implement it if you feel its convenient to your project.
Some lsp servers use a command line "--log-file" flag for this purpose or maybe some "workspace/configuration" could be used too.

Oh, and JJ is an acronym for "stupid me", a useful trick to mark code I messed with because "JJ" is seldom found in code bases. (Incidentally those are my initials).

File: server/lib/JJ.pm (hunk 1 above)
Comment (mixio) [reply server/lib/JJ.pm:24]: Noted

File: server/lib/JJ.pm (hunk 2 above)
Comment (mixio) [reply server/lib/JJ.pm:20]: Noted

File: server/lib/PLS/Server.pm (hunk 3 above)
Comment (mixio) [reply server/lib/PLS/Server.pm:139]: Client

File: server/lib/PLS/Server.pm (hunk 4 above)
Comment (mixio) [reply server/lib/PLS/Server.pm:218]: Server.
The underscore helps in reading the log files:
S_: {... message from the server }
_C: {... message from the client }

File: server/lib/PLS/Server/Method/TextDocument.pm (hunk 5 above)
Comment (mixio) [reply server/lib/PLS/Server/Method/TextDocument.pm:130]: Ignore

File: server/lib/PLS/Server/Request/Exit.pm (hunk 6 above)
Comment (mixio) [reply server/lib/PLS/Server/Request/Exit.pm:23]: Please implement as required

File: server/lib/PLS/Server/Request/Shutdown.pm (hunk 7 above)
Comment (mixio) [reply server/lib/PLS/Server/Request/Shutdown.pm:25]: Please implement as required

File: server/t/packets/shutdown.json (hunk 8 above)
Comment (mixio) [reply server/t/packets/shutdown.json:3]: Please implement as required

**Comment by FractalBoy** [comment]
Superceded by #88
